packages/gatsby-source-graphql/README.md

```diff
@@ -176,5 +176,162 @@ module.exports = {
 }
 ```
 
+# Performance tuning
+
+By default, `gatsby-source-graphql` executes each query in a separate network request.
+But the plugin also supports query batching to improve query running performance
+
+**Caveat**: Batching is only possible for queries starting approximately at the same time. In other words
+it is bounded by the number of parallel GraphQL queries executed by Gatsby (by default it is **4**).
+
+Fortunately we can increase the number of queries executed in parallel by setting [environment variable](https://gatsby.dev/env-vars)
+`GATSBY_EXPERIMENTAL_QUERY_CONCURRENCY` to a higher value and setting `batch` option of the plugin
+to `true`.
+
+Example:
+
+```
+cross-env GATSBY_EXPERIMENTAL_QUERY_CONCURRENCY=20 gatsby develop
+```
+
+With plugin config:
+
+```js
+const fs = require("fs")
+const { buildSchema, buildClientSchema } = require("graphql")
+
+module.exports = {
+  plugins: [
+    {
+      resolve: "gatsby-source-graphql",
+      options: {
+        typeName: "SWAPI",
+        fieldName: "swapi",
+        url: "https://api.graphcms.com/simple/v1/swapi",
+        batch: true,
+      },
+    },
+  ],
+}
+```
+
+By default, the plugin batches up to 5 queries. You can override this by passing
+`dataLoaderOptions`:
+
+```js
+const fs = require("fs")
+const { buildSchema, buildClientSchema } = require("graphql")
+
+module.exports = {
+  plugins: [
+    {
+      resolve: "gatsby-source-graphql",
+      options: {
+        typeName: "SWAPI",
+        fieldName: "swapi",
+        url: "https://api.graphcms.com/simple/v1/swapi",
+        batch: true,
+        // See https://github.com/graphql/dataloader#new-dataloaderbatchloadfn--options
+        // for a full list of DataLoader options
+        dataLoaderOptions: {
+          maxBatchSize: 10,
+        },
+      },
+    },
+  ],
+}
+```
+
+Having 20 parallel queries with 5 queries per batch means we are still running 4 batches
+in parallel.
+
+Each project is unique so try tuning those two variables and see what works best for you.
+We've seen up to 5-10x speed-up for some setups.
+
+### How batching works
+
+Under the hood `gatsby-source-graphql` uses [DataLoader](https://github.com/graphql/dataloader)
+for query batching. It merges all queries from a batch to a single query that gets sent to the
+server in a single network request.
+
+Consider following example:
+
+```js
+{
+  query: `query(id: Int!) {
+    node(id: $id) {
+      foo
+    }
+  }`,
+  variables: { id: 1 },
+}
+```
+
+```js
```

Comment:
```suggestion
```graphql
```

Reply:
Also a JS example (graphql query is a part of JS object here)